HiDPI Skin Scaling

daschuer opened the pull request:
This is the PR original started by @devananda to allow to build skins using @2x suffixed images for HiDpi screens. 

I have picked it up, because we have an increasing amount of reports complaining about missing HiDPI support on Win and Linux. Mac has already "some" HiDpi support by the OS, for font rendering, and upscaling. so there seams to be less pressure.  

I have now completed his work with floating point scaling for almost all skin widgets.
For now, I have kept the checkbox in preferences for double only. I consider fractional scaling as an experimental option, that can be enabled by editing the cfg file. `ScaleFactor 0.5`
This is usefull for example to test big skins on small displays. 
It is fun to test exotic scaling values B-)

I have failed to scale the combobox which is heavy styled by a style sheet. 
Qt4 has no nice interface to hock between style sheet ans widget painting. 
So other styled elements are effected as well. 
I think a complete solution for this requires Qt5. 

I have no HiDPI screen, so I cannot decide if this is already usable. 
Assuming this does not break 1x scaling, this is a candidate to include it into 2.1 marked as experimental.

In any case this is a step in a right direction, because it allows high resolution cover arts and waveforms.

AevaOnline committed:
Double the scale of all widgets

Modifies legacyskinparser.cpp to double the scale of all widgets,
following tips found here:
      http://mixxx.org/forums/viewtopic.php?f=8&t=7098&sid=0f5a2a95eb716ebcd9664dbd76623620#p24281

AevaOnline committed:
Introduce DoubleWidgetSize config option

Introduce a new Interface config option that doubles the expected sizes
of all widgets.

NOTE: this only renders properly if the widget graphics (PNG files) are
actually 2x the expected size. However, it is not loading the "@2x"
files properly, so this is not working yet.

AevaOnline committed:
WPixMapStore loads alternate image file names

AevaOnline committed:
Reload skin when toggling DoubleWidgetSize option

AevaOnline committed:
Fall back to normal image file name if "@2x" alt file doesn't exist

daschuer committed:
allow floating point scale factors

daschuer committed:
move scaleFactor to SkinContext

daschuer committed:
make m_loader not null and remove now unreachable code

daschuer committed:
use const refferences and override attribute

daschuer committed:
use scaleFactor when loading images

daschuer committed:
scale "pos"

daschuer committed:
Scale Fonts

daschuer committed:
Scale svg images

daschuer committed:
Scale woverview

daschuer committed:
Add skale factor to lookup key

daschuer committed:
Fix re-loading skin after changing from threestate

daschuer committed:
Addes ScaleFactor to the waveform renderers

daschuer committed:
Fix lable scaling in Woverview

daschuer committed:
Fix knob scaling

daschuer commented:
Deere @0.5x
![bildschirmfoto vom 2017-03-05 01 55 48](https://cloud.githubusercontent.com/assets/1777442/23583515/fc1e839a-0146-11e7-9aaf-3c22d4dac271.png)

daschuer commented:
@devananda: I have removed the upscaled @2x versions, because the on-the-fly generated version are quite similar. To improve the @2x quality some @2x images with removed aliasing would be nice. 
An other option would be to switch to svg versions. Unfortunately this comes with a penalty of skin loading time.

Be-ing commented on src/preferences/dialog/dlgprefcontrolsdlg.ui:
```diff
@@ -489,6 +489,23 @@
       </widget>
      </item>
+     <item row="12" column="1" colspan="2">
+      <widget class="QCheckBox" name="checkBoxDoubleWidgetSize">
```
Maybe this should be a QDoubleSpinBox?

daschuer replied:
For now this is a checkbox by design, because we "only" have the request for double the size. 
And I want to avoid to have to much degrees of freedom for now. 

If you thing it is useful to allow free scaling, we can consider to go for it. Today almost all is calculated on pixel base so that 1x and 2x works nice and fractional scaling is kind of blurry.
It now supports @2x suffix to avoid this issue.

Be-ing replied:
I know other ratios are common on smartphone screens, but Mixxx does not run on those. I am not sure about laptops.

Be-ing replied:
The desired scale factor may vary depending on screen resolution, physical screen size, and personal taste, so I think it would be good to make it more flexible.

daschuer committed:
Try to fix clang build

Be-ing commented:
Build error with gcc-6.3.1-1.fc25.x86_64:
```g++ -o lin64_build/skin/skincontext.o -c -std=c++11 -pipe -Wall -Wextra -g -O3 -ffast-math -funroll-loops -fomit-frame-pointer -march=native -pthread -Dx86_64 -DMIXXX_BUILD_DEBUG -D__LINUX__ -D__UNIX__ -DSETTINGS_PATH=\".mixxx/\" -DSETTINGS_FILE=\"mixxx.cfg\" -DUNIX_SHARE_PATH=\"/home/be/local/share/mixxx\" -DUNIX_LIB_PATH=\"/home/be/local/lib/mixxx\" -D__PORTAUDIO__ -DQT_TABLET_SUPPORT -DQT_SHARED -DQT_CORE_LIB -DQT_GUI_LIB -DQT_OPENGL_LIB -DQT_XML_LIB -DQT_SVG_LIB -DQT_SQL_LIB -DQT_SCRIPT_LIB -DQT_NETWORK_LIB -DQT_SHARED -D__SNDFILE__ -D__MAD__ -D__HID__ -D__BULK__ -D__VINYLCONTROL__ -D__BROADCAST__ -D__OPUS__ -D__VAMP__ -Dkiss_fft_scalar=double -DHAVE_FFTW3 -D__SQLITE3__ -D__BATTERY__ -Ilin64_build -Isrc -Ilib/soundtouch-1.9.2 -Ilib/replaygain -Ilib/libebur128-1.1.0/ebur128 -I/usr/include/QtOpenGL -I/usr/include/QtXml -I/usr/include/QtSvg -I/usr/include/QtSql -I/usr/include/QtNetwork -I/usr/include/QtTest -I/usr/include/QtScriptTools -I/usr/include/QtGui -I/usr/include/QtScript -I/usr/include/QtCore -Ilib/gtest-1.7.0/include -Ilib/fidlib-0.9.10 -I/usr/include/taglib -Ilib/qtscript-bytearray -Ilib/reverb -Ilib/portaudio -I/usr/include/libusb-1.0 -Ilib/hidapi-0.8.0-rc1/hidapi -Ilib/xwax -Ilib/scratchlib -I/usr/include/opus -I/usr/include/libupower-glib -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include src/skin/skincontext.cpp
src/skin/skincontext.cpp: In member function ‘int SkinContext::scaleToWidgetSize(QString&) const’:
src/skin/skincontext.cpp:268:60: error: ‘round’ was not declared in this scope
         return static_cast<int>(round(dSize * m_scaleFactor));
                                                            ^
scons: *** [lin64_build/skin/skincontext.o] Error 1
scons: building terminated because of errors.```

daschuer committed:
Add missing include

Be-ing commented:
I tried this on my 1366 x 768 laptop screen and it seems to work as advertised. Setting ScaleFactor to 2.0 made the GUI very laggy for me. I'm guessing that's because of my old integrated GPU, so I'm guessing it wouldn't be an issue on a newer computer that has a screen resolution high enough to need this feature. Scaling down to 0.5 did not make the GUI lag.

Be-ing commented:
> I have failed to scale the combobox which is heavy styled by a style sheet.
Qt4 has no nice interface to hock between style sheet ans widget painting.
So other styled elements are effected as well.
I think a complete solution for this requires Qt5.

Unfortunately the QDoubleSpinBox subclass introduced in #1187 suffers the same problem. :/

esbrandt commented:
Thanks for working on it.

Tested the current state of the PR on macOS with latest master.

  * Pushbutton images look jagged, even if unscaled. Deere uses *.svg, Latenight *.png see screenshot below.
    ![deere 1 hidpi default-1](https://cloud.githubusercontent.com/assets/4525897/23836035/b98731ae-0771-11e7-9cf3-32aa2735a163.png)
    ![deere 2 hidpi](https://cloud.githubusercontent.com/assets/4525897/23836036/be5fa79c-0771-11e7-83a1-005c8be214c9.png)
    ![latenight 1 hidpi default](https://cloud.githubusercontent.com/assets/4525897/23836064/3b18b36e-0772-11e7-9527-3c304e247447.png)
    ![latenight 2 hidpi](https://cloud.githubusercontent.com/assets/4525897/23836046/f2f994ea-0771-11e7-9b63-d57091892fe1.png)
  * When  switching from a skin that supports color schemes, back to another skin, somehow the color scheme processor does not reset. In the example below, i changed to Dark Metal, selected another color scheme, than switched to Latenight. Some Pushbuttons, sliders and background colors are tinted.
    ![mixxx-117](https://cloud.githubusercontent.com/assets/4525897/23836083/b6734e2a-0772-11e7-81ed-dbf6b5de1361.png)
  * Naming of the preferences option. Lend the wording from Windows 10, here as i find it pretty spot on, and usable even if we decide to support other resolutions then x2 (e.g. by a slider)
   ``HiDPI / Retina scaling: Change the size of Text, buttons, and other items``
   Having the HiDPI/Retina buzzwords helps to catch the eye, and describe in user centric terms what the option does.
  * Just for the case someone comes up with a HiDPI skin, and the user has the x2 option enabled, is there a way to handle that? Like scanning the skin manifest for a (not yet existing) ``HiDPI`` node, determined the resolution, and defaulting to x1?

Be-ing commented:
> Pushbutton images look jagged, even if unscaled

I noticed this too on GNU/Linux (X11).

> HiDPI / Retina scaling: Change the size of text, buttons, and other items

SGTM

> Just for the case someone comes up with a HiDPI skin, and the user has the x2 option enabled, is there a way to handle that? Like scanning the skin manifest for a (not yet existing) HiDPI node, determined the resolution, and defaulting to x1?

I don't think we should have skins that only work on HiDPI displays.

esbrandt commented:
> I don't think we should have skins that only work on HiDPI displays.

Why not, doesn't have to be the default...
I do not have a HiDPI screen, but imagine developing a x0.5 skin on a HiDPI screen is no fun. I would rather take any good skin than booting it out. But this is a non-issue, if we allow for scaling x0.5.

Be-ing commented:
> Why not, doesn't have to be the default...

Because of the maintenance burden. We should not go back to having different skins for different resolutions.

> I do not have a HiDPI screen, but imagine developing a x0.5 skin on a HiDPI screen is no fun. I would rather take any good skin than booting it out. But this is a non-issue, if we allow for scaling x0.5.

I think scaling is a better solution than maintaining separate designs.

daschuer commented:
>> I don't think we should have skins that only work on HiDPI displays.

> Why not, doesn't have to be the default...
> I do not have a HiDPI screen, but imagine developing a x0.5 skin on a HiDPI screen is no fun. I would rather take any good skin than booting it out. But this is a non-issue, if we allow for scaling x0.5.

I think we should allow this. We could allow to develop a skin using @2x suffixed images only. If a user starts Mixxx with default settings (1x) on a non HiDPI screen. Theses images should be automatically scaled down to 0.5x

daschuer committed:
Reset coloer schema if there is no "Schemas" section found in skin.

daschuer commented:
The latest commit fixes the color schema issue and also the "jagged" issue. 

The remaining issue that the play/pause button looks odd, compared to the CUE button is because one is a vector text button and the other is simply an upscaled image.
This can be fixed in the skins by replacing the bitmap images by vector graphic or providing a @2x suffixed alternative.

daschuer committed:
Fall back to @2x or @3x image if no image if found in the desired resolution

daschuer commented:
@esbrandt: Should I change the preference option to a slider, or is the check-box option sufficient for the first version? If we decide for a slider, what should be the range? Free 0.5 to 3 or snappy to some predefined and tested scale values?

Be-ing commented:
I don't understand the need for @2x or @3x images when we can use vector images.

daschuer commented:
> I don't understand the need for @2x or @3x images when we can use vector images.

It is just to have the freedom to use bitmaps. Bitmaps are loading much faster and it may happen that we have no vector sourced from some images.

Be-ing commented:
I tested again and confirm that the jagged edges with a 1.0 scale factor have been fixed.

Be-ing commented:
Actually, the jagged edges are not entirely fixed.
master:
![smooth](https://cloud.githubusercontent.com/assets/9455094/23843163/d7bc616a-0787-11e7-9dfe-3dc5ecbde8fe.png)
this PR:
![jagged-edges](https://cloud.githubusercontent.com/assets/9455094/23843164/d7c0d3da-0787-11e7-9f5c-c32da718e05c.png)

esbrandt commented:
@daschuer 
Any reason for not only allowing predefined values, like 0.5, 1, 2, 3(4K), 4(5K)?

> Actually, the jagged edges are not entirely fixed without scaling:

Can confirm

daschuer committed:
Use a combobox for setting the scale factor, tweak the dlg preferences and add a scalFactor Auto checkboc for detecting system wide scale factors.

daschuer committed:
Don't prescale stretched images

daschuer commented:
The jagged edges are gone now and we have the proposed scale factors in a combobox.

esbrandt commented on src/preferences/dialog/dlgprefcontrols.cpp:
```diff
@@ -298,6 +298,46 @@ DlgPrefControls::DlgPrefControls(QWidget * parent, MixxxMainWindow * mixxx,
+            ConfigKey("[Config]", "ScaleFactor"), 1.0);
+
+    //: Entry of the HiDPI scale combo box. %1 is the scale factor-
+    comboBoxScaleFactor->addItem(QString(tr("x %1")).arg(0.5), 0.5);
```
A scale factor is usable, using percent would be more consistent with the existing options.
E.g. the waveform zoom level is given in percent.

A tooltip would be a nice touch (``Change the size of text, buttons, and other items.``), as well as for the Auto-Scaling checkbox.

Nit: Moving the new preference option below the Locale option, so the three expanding comboboxes are layered, makes it more easy on the eyes IMO.

RE: the combobox scaling issue, anything the skin designer can do with the combobox styling to allow proper scaling?

Be-ing commented:
Maybe there is a way to work around the scaling issue with WEffectSelector by reimplementing some QWidget function? Just a guess.

daschuer commented:
I have actually started to reimplement hack around the combo box painting functions. But it turns out to be a pile of work and/or quite hackie, sins it is a complex widget. Facing on the other hand a well designed scaling feature in Qt 5, it is questionable if this would be worth the effort, doing this for a single release. 

So we come back to the main question: Is this worth to be part of 2.1 even if it not perfect? I would say yes, if there are no regressions for the 100 % case. I would say yes, because we have at least one user (see Launchpad bug) who is able to  use Mixxx now,  won't using it without.
We may mark this feature as preliminary, or so.

Be-ing commented:
Alternatively we could provide a Qt5 build or say to build it yourself with Qt5 and warn that it is not well tested.

Be-ing commented:
Generally speaking, I'd like to not settle for features that only kinda work. IMO it gives the impression that Mixxx is a half-hearted effort. I'd rather instruct users how to build with Qt5 and/or provide Qt5 builds with a disclaimer than than ship a version where effect names are too small to read.

daschuer commented:
For most users it is no alternative to build a untested Qt 5 version at home. For now, we have the alternative to ship Mixxx 2.1 without this PR and exclude all HiDPI users, or include this and allow them to use the default installation with some issues.

Yes, it is odd to have small effect names, but it is even better than have everything small. Effects are not a blocker feature, they can still use them, or fall back to a skin with out a combo box selector.

If we mark this feature explicit as "half-hearted" alias "experimental" it will not this a bad light on whole Mixxx.

Be-ing commented:
> If we mark this feature explicit as "half-hearted" alias "experimental" it will not this a bad light on whole Mixxx.

I think that's a good idea. I'll think about what words to use for the disclaimer.

daschuer committed:
Use percentage for HiDpi scaling, Add "(Experimental)" to non 100 % values

daschuer committed:
Improve preferences order and add tooltips

Be-ing commented:
How about specifying an @2x style sheet?

daschuer committed:
implement auto HiDpi scaling using gsettings

daschuer committed:
Load style@2x.qss or style@3x.qss in addition to the prefix less file

daschuer commented:
Now it scales automatically on a Linux desktops featuring gsettings. 
I have also added the stylesheet "@2x" and "@3x"

I have not tried yet to fix the combobox scaling, but it should work now using the postfixed stylesheets.

Interestingly the combobox scales already with the fonts using Windows.

Be-ing commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1789,6 +1789,31 @@ QString LegacySkinParser::getStyleFromNode(const QDomNode& node) {
+                                               fileBytes.length()));
+            }
+        } else if (scaleFactor >= 2) {
+            // Try to load with @3x suffix
```
typo in comment, should say `@2x`

Be-ing commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1789,6 +1789,31 @@ QString LegacySkinParser::getStyleFromNode(const QDomNode& node) {
+                style.prepend(QString::fromLocal8Bit(fileBytes.constData(),
+                                               fileBytes.length()));
+            }
+        } else if (scaleFactor >= 2) {
```
I'm wondering if maybe the threshold for this should be lower. Maybe > 1.5?

daschuer replied:
I have adopted this form the Gnome/Unity Scaling slider. 
It is one floating point sider for the fonts and widgets. The widgets are doubled once you reach the 2.0 threshold. 
This means, once we are at Qt 5, the native widgets will also scale at this point, Which may require some style tweaks, at this threshold. 

I will add a comment.

Be-ing commented:
> I have not tried yet to fix the combobox scaling, but it should work now using the postfixed stylesheets.

Hopefully it will be as simple as doubling the pixel sizes for the comboboxes...

> Interestingly the combobox scales already with the fonts using Windows.

That is interesting. What about on Mac OS X? @esbrandt can you test that?

daschuer commented:
Unfortunately the combo box handle icon is not scaled on windows. ...
For my feeling this is a topic of continuously tweaking once this is merged. 

I still do not know how to detect the scaling settings from Windows and Mac. 
Can anyone jump in here? 

Mixxx 2.0 runs on Mac OS as a non Retina aware app. This means it is upscaled by the OS (The reason we have no complains by Mac OS users.) It results into sharp fonts, but blurry cover arts and waveforms. 
Once this patch is merged, we may have to tell Mac OS that Mixxx does not need to be scaled. 
I am curios if the result is better than ...  I am afraid not. 
So it is finally a good idea to have this feature marked as "experimental".

Be-ing commented:
> Unfortunately the combo box handle icon is not scaled on windows. ...

That's annoying, but not that big of a deal. What is important is that the text is readable.

daschuer committed:
fix and add comments

daschuer commented:
Fixed, what else do we need before merge?

Be-ing commented:
I tested this again and there don't seem to be any regressions at 100% scale. I played a little with styling WEffectSelector and WBeatSpinBox and found that the most important style to specify is font-size. If that would be doubled again on Windows, we may need to disable loading the @2x/@3x stylesheets for Windows. LGTM for merge. We'll need to get feedback in wider beta testing to work out the details of this the best we can.

Be-ing commented:
Should this be excluded from Qt5 builds with ifdefs?

daschuer committed:
Fix a segfault with classic Shade skin

daschuer committed:
Add font skaling for WEffectSelector and define a fixed pixel size font

daschuer committed:
Remove point size font scaling. They are already scaled by the OS

Be-ing commented:
It looks like there is a merge conflict now, probably because of #1219. That was a tiny PR, so it should be easy to resolve.

Be-ing commented:
The font scaling works well for WEffectSelector. Maybe we can remove the @2x/@3x stylesheet support.

I noticed the library fonts are not scaled.

daschuer commented:
I think I have discovered the rule: 
You can set the font size by pixel (px) or by point (pt) size. 
The point size fond depends on the Screen DPI, which changes with the OS Desktop scaling (verified with Ubuntu Unity and Windows 10)
Original the WEffectSelector has no font size set, and defaults to the default point sized font. 
I have changed it now to a px size fond, and removed all mixxx scaling from point size fonts to avoid double scaling. 

The style @2x hack could be used to tweak for example margin and the combobox handle. 
Do you think keeping this will introduce future issues? If we do not make use of it for 2.1 I should remove or disable it.

There is still an issue with the Track table header. It seams that this is a pixel size font, because it is not OS scaled. Do you know where it is define? This has to be a point size font. 

We can also enable the Auto feature for Windows by query the widget DPI.

esbrandt commented:
> track table header

In Deere: https://github.com/mixxxdj/mixxx/blob/master/res/skins/Deere/style.qss#L178

daschuer commented:
Ups .. that easy :-) Thank you!

daschuer committed:
tweak font size for scalable

daschuer committed:
disable @2x styles for now, until we know better.

esbrandt reviewed (commented):
Originally, we had all font sizes given in pt, but switched to px sizes. I can not remember the details, but there was possible a cross-platform issue. 

So we need to test if this change does not have unwanted consequences.

Edit: Found 98eec1e6a2ecd5be64078d89164a53b56bd53719

Be-ing commented:
> There is still an issue with the Track table header

It's not just the table headers; it's also the text in the table cells that is not scaled.

daschuer commented:
We have to distinguish between the skin and the library region. In the skin region, we have to use pixel size fonds in our custom widgets, to not have cross platform issues and to make the new scaling option from this PR work. The text on buttons should not resize with the OSs font setting. 

In the library area, we have the "Library Font" preference option to scale this area separate. This is a point size font, which is scaled by the OSs DPI setting. The same is true for other dialogs like preferences. 

If we now look at the header bar of the track table we have two options: 
Use pixel size fonts and override the header bar font code to resize them annual. Or just use point size fonts and rely on the OSs font size settings.  
In this case it is convenient that the header follows the OSs setting like the table content.

daschuer committed:
Merge remote-tracking branch 'upstream/master' into LateNight

Conflicts:
	src/widget/weffectselector.cpp

daschuer committed:
Added autoHiDpi for windows

daschuer commented:
now with auto HiDpi for windows as well

daschuer commented:
Ready for merge ?!

Be-ing commented:
How does this behave with Qt5? Should the scale factor be hardcoded as 1.0 with Qt5 and the option hidden from the preferences?

daschuer committed:
Merge remote-tracking branch 'upstream/master' into LateNight

Conflicts:
	src/preferences/dialog/dlgprefcontrols.h
	src/preferences/dialog/dlgprefcontrolsdlg.ui

daschuer commented:
> How does this behave with Qt5? Should the scale factor be hardcoded as 1.0 with Qt5 and the option hidden from the preferences?

I have not tested it with QT5 and I would like to avoid the work on QT5 for now since my last effort produces a crashing Mixxx. So it would be OK for me to hide the option for now.

daschuer committed:
Disaple Scaling code for QT5 until we have time to test it

daschuer committed:
Adopt a new Auto HiDPI scaling value on update

daschuer commented:
Done.

Be-ing commented:
LGTM! Thank you for making Mixxx usable on newer computers. After this is merged I will adopt the font scaling strategy used for WEffectSelector in this PR for WBeatSpinBox in #1187.

tristan-k commented:
How do I test this? I'm using a surface device (3000x2000 resolution). I downloaded the latest build from http://downloads.mixxx.org/builds/master/release/mixxx-2.1.0-alpha-pre-master-git6233-release-x64.msi, but I can't find any option for high dpi displays in the preferences.

daschuer commented:
You should try this: 
https://ci.appveyor.com/api/buildjobs/hakiak1wqjuomd32/artifacts/mixxx-2.1.0-alpha-pre-master-PR1204-git6234-release-x64.msi

Please check if Mixxx scales with Windows when changing from 100 to 200 % in the windows preferences and Auto Scaling enabled (default). 

Could you provide a screen shot? This was developed on a SD screen.

tristan-k commented:
Sure. After a quick look it seems to scale properly.

![mixxx_highdpi](https://cloud.githubusercontent.com/assets/732025/24502585/b9f4080c-154e-11e7-9fe6-2d6f826d567a.PNG)

daschuer commented:
Thank you! For testing and review and @devananda for a solid start! This is finally even better than I had hoped initially.
